diff --git a/tests/environment.nix b/tests/environment.nix
index eb517883c..2b0380ed7 100644
--- a/tests/environment.nix
+++ b/tests/environment.nix
@@ -1,12 +1,16 @@
 { var, value }:
 
-derivation {
+with import ./config.nix;
+
+mkDerivation {
   name = "test";
-  system = builtins.currentSystem;
-  builder = "/bin/sh";
-  args = [ "-c" ''
-    echo ${var} = "''$${var}"; echo "''$${var}" > "$out"'' ];
+  buildCommand = ''
+    echo ${var} = "''$${var}"
+    echo -n "''$${var}" > "$out"
+  '';
+
   impureEnvVars = [ var ];
+
   outputHashAlgo = "sha256";
-  outputHash = builtins.hashString "sha256" "${value}\n";
+  outputHash = builtins.hashString "sha256" value;
 }
