diff --git a/src/nix/search.cc b/src/nix/search.cc
index 4fa1e7837..7ca0421b2 100644
--- a/src/nix/search.cc
+++ b/src/nix/search.cc
@@ -9,6 +9,8 @@
 #include "eval-cache.hh"
 #include "attr-path.hh"
 #include "hilite.hh"
+#include "fmt.hh"
+#include "value-to-json.hh"
 
 #include <regex>
 #include <fstream>
@@ -153,10 +155,13 @@ struct CmdSearch : InstallableCommand, MixJSON
                     {
                         results++;
                         if (json) {
+                            PathSet context;
+
                             (*jsonOut)[attrPath2] = {
                                 {"pname", name.name},
                                 {"version", name.version},
                                 {"description", description},
+                                {"meta", printValueAsJSON(*state, true, aMeta->forceValue(), noPos, context, false)}
                             };
                         } else {
                             auto name2 = hiliteMatches(name.name, nameMatches, ANSI_GREEN, "\e[0;2m");
