diff --git a/src/nix/search.cc b/src/nix/search.cc
index c8d0b9e96..1510b2955 100644
--- a/src/nix/search.cc
+++ b/src/nix/search.cc
@@ -11,6 +11,8 @@
 #include "attr-path.hh"
 #include "hilite.hh"
 #include "strings-inline.hh"
+#include "fmt.hh"
+#include "value-to-json.hh"
 
 #include <regex>
 #include <fstream>
@@ -155,10 +157,13 @@ struct CmdSearch : InstallableValueCommand, MixJSON
                     {
                         results++;
                         if (json) {
+                            NixStringContext context;
+
                             (*jsonOut)[attrPath2] = {
                                 {"pname", name.name},
                                 {"version", name.version},
                                 {"description", description},
+                                {"meta", printValueAsJSON(*state, true, aMeta->forceValue(), noPos, context, false)}
                             };
                         } else {
                             auto name2 = hiliteMatches(name.name, nameMatches, ANSI_GREEN, "\e[0;2m");
