diff --git a/src/nix/search.cc b/src/nix/search.cc
index e9307342c..9f293fff8 100644
--- a/src/nix/search.cc
+++ b/src/nix/search.cc
@@ -10,6 +10,7 @@
 #include "eval-cache.hh"
 #include "attr-path.hh"
 #include "hilite.hh"
+#include "value-to-json.hh"
 
 #include <regex>
 #include <fstream>
@@ -130,6 +131,11 @@ struct CmdSearch : InstallableCommand, MixJSON
                             jsonElem.attr("pname", name.name);
                             jsonElem.attr("version", name.version);
                             jsonElem.attr("description", description);
+
+                            auto jsonMeta = jsonElem.placeholder("meta");
+
+                            PathSet context;
+                            printValueAsJSON(*state, true, aMeta->forceValue(), noPos, jsonMeta, context);
                         } else {
                             auto name2 = hiliteMatches(name.name, std::move(nameMatches), ANSI_GREEN, "\e[0;2m");
                             if (results > 1) logger->cout("");
