diff --git a/src/nix/search.cc b/src/nix/search.cc
index dac60ceba..b72dea4bd 100644
--- a/src/nix/search.cc
+++ b/src/nix/search.cc
@@ -11,6 +11,8 @@
 #include "nix/expr/attr-path.hh"
 #include "nix/util/hilite.hh"
 #include "nix/util/strings-inline.hh"
+#include "nix/util/fmt.hh"
+#include "nix/expr/value-to-json.hh"

 #include <regex>
 #include <fstream>
@@ -147,10 +149,13 @@ struct CmdSearch : InstallableValueCommand, MixJSON
                     if (found) {
                         results++;
                         if (json) {
+                            NixStringContext context;
+
                             (*jsonOut)[attrPathStr] = {
                                 {"pname", name.name},
                                 {"version", name.version},
                                 {"description", description},
+                                {"meta", printValueAsJSON(*state, true, aMeta->forceValue(), noPos, context, false)}
                             };
                         } else {
                             if (results > 1)
