diff --git a/src/nix/search.cc b/src/nix/search.cc
index a27891c93..d033df4fe 100644
--- a/src/nix/search.cc
+++ b/src/nix/search.cc
@@ -11,6 +11,8 @@
 #include "nix/expr/attr-path.hh"
 #include "nix/util/hilite.hh"
 #include "nix/util/strings-inline.hh"
+#include "nix/util/fmt.hh"
+#include "nix/expr/value-to-json.hh"
 
 #include <regex>
 #include <fstream>
@@ -155,10 +157,13 @@ struct CmdSearch : InstallableValueCommand, MixJSON
                     {
                         results++;
                         if (json) {
+                            NixStringContext context;
+
                             (*jsonOut)[attrPath2] = {
                                 {"pname", name.name},
                                 {"version", name.version},
                                 {"description", description},
+                                {"meta", printValueAsJSON(*state, true, aMeta->forceValue(), noPos, context, false)}
                             };
                         } else {
                             if (results > 1) logger->cout("");
